src/EFRepository.Generator/ExtensionMethodGenerator.cs

```diff
@@ -1,237 +1,388 @@
+	{
+		if (context.SyntaxReceiver is not SyntaxReceiver receiver)
+			return;
+
+		var options = ((CSharpCompilation)context.Compilation).SyntaxTrees[0].Options as CSharpParseOptions;
+		var compilation = context.Compilation;
+
+		var dbContextSymbol = compilation.GetTypeByMetadataName("Microsoft.EntityFrameworkCore.DbContext");
+
+		if (dbContextSymbol == null)
+			dbContextSymbol = compilation.GetTypeByMetadataName("Microsoft.EntityFramework.DbContext");
+
+		if (dbContextSymbol == null)
+			return;
+
+		var dbSetSymbol = compilation.GetTypeByMetadataName("Microsoft.EntityFrameworkCore.DbSet`1");
+
+		if (dbSetSymbol == null)
+			dbSetSymbol = compilation.GetTypeByMetadataName("Microsoft.EntityFramework.DbSet");
+
+		if (dbSetSymbol == null)
+			return;
+
+		var dbContexts = receiver.Classes.Select(c =>
+		{
+			var sourceModel = compilation.GetSemanticModel(c.SyntaxTree);
+			var sourceSymbol = sourceModel.GetDeclaredSymbol(c);
+
+			if (sourceModel == null || sourceSymbol == null)
+				return null;
+
+			return new SourceRecord
+			{
+				Model = sourceModel,
+				Symbol = sourceSymbol,
+				Syntax = c
+			};
+		})
+		.Where(c => c != null)
+		.Where(c =>
+			c != null &&
+			c.Symbol != null &&
+			c.Symbol.ContainingSymbol.Equals(c.Symbol.ContainingNamespace, SymbolEqualityComparer.Default) &&
+			c.Symbol.BaseType != null && c.Symbol.BaseType.Equals(dbContextSymbol, SymbolEqualityComparer.Default));
+
+		// TODO: Need to check that this is actually a DbContext class
+
+		var extensions = dbContexts.SelectMany(dbc => ProcessDbContext(dbc, dbSetSymbol, compilation));
+
+		foreach (var (FileName, Content) in extensions)
+		{
+			context.AddSource(FileName, SourceText.From(Content, Encoding.UTF8));
+		}
+	}
+
+	protected IEnumerable<(string FileName, string Content)> ProcessDbContext(SourceRecord record, INamedTypeSymbol dbSetSymbol, Compilation compilation)
+	{
+		var dbSets = record.Symbol.GetMembers()
+		.Where(m => !m.IsStatic && !m.IsAbstract && !m.IsImplicitlyDeclared
+					&& m.Kind is SymbolKind.Property && m.IsDefinition)
+		.Cast<IPropertySymbol>()
+		.Where(p => !p.IsReadOnly && !p.IsExtern)
+		.Where(p =>
+		{
+			return p.Type.OriginalDefinition.Equals(dbSetSymbol, SymbolEqualityComparer.Default);
+		});
+
+		var list = new List<(string FileName, string Content)>();
+
+		foreach (var dbSet in dbSets)
+		{
+			var genericType = ((INamedTypeSymbol)dbSet.Type).TypeArguments[0];
+
+			if (genericType == null)
+				continue;
+
+			var sourceCode = BuildExtensionMethod((INamedTypeSymbol)genericType, compilation);
+
+			list.Add(sourceCode);
+		}
+
+		return list;
+	}
+
+	protected (string FileName, string Content) BuildExtensionMethod(INamedTypeSymbol dbSetClass, Compilation compilation)
+	{
+		string fileName = $"{dbSetClass.Name}.EFExtensions.g.cs";
+
+		var builder = new StringBuilder(
+$@"using System;
+using System.Linq;
+using Microsoft.EntityFrameworkCore;
+
+namespace {dbSetClass.ContainingNamespace};

```

Comment:
We cannot use this feature here since this project also targets .NET Framework, which doesn't support the new namespace syntax. This also presupposes that the user is running version 10 of the language, which may also not be the case.

Reply:
Reverted the file-scoped namespacing